xds/src/test/java/io/grpc/xds/XdsClientFederationTest.java

```diff
@@ -0,0 +1,153 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static org.mockito.ArgumentMatchers.isA;
+import static org.mockito.Mockito.timeout;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.grpc.xds.XdsClient.ResourceWatcher;
+import io.grpc.xds.XdsListenerResource.LdsUpdate;
+import java.util.Collections;
+import java.util.Map;
+import java.util.UUID;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.mockito.Mock;
+import org.mockito.MockitoAnnotations;
+
+/**
+ * Tests for xDS control plane federation scenarios.
+ */
+@RunWith(JUnit4.class)
+public class XdsClientFederationTest {
+
+  @Mock
+  private ResourceWatcher<LdsUpdate> mockDirectPathWatcher;
+
+  @Mock
+  private ResourceWatcher<LdsUpdate> mockWatcher;
+
+  private static final String SERVER_LISTENER_TEMPLATE_NO_REPLACEMENT =
+      "grpc/server?udpa.resource.listening_address=";
+
+  @Rule
+  public ControlPlaneRule trafficdirector = new ControlPlaneRule("test-server");
+
+  @Rule
+  public ControlPlaneRule directpathPa = new ControlPlaneRule(
+      "xdstp://server-one/envoy.config.listener.v3.Listener/test-server");
+
+  private XdsClient xdsClient;
+  private boolean originalFederationStatus;
+
+  @Before
+  public void setUp() throws XdsInitializationException {
+    MockitoAnnotations.initMocks(this);
+
+    originalFederationStatus = BootstrapperImpl.enableFederation;
+    BootstrapperImpl.enableFederation = true;
+
+    SharedXdsClientPoolProvider clientPoolProvider = new SharedXdsClientPoolProvider();
+    clientPoolProvider.setBootstrapOverride(defaultBootstrapOverride());
+    xdsClient = clientPoolProvider.getOrCreate().getObject();
+  }
+
+  @After
+  public void cleanUp() throws InterruptedException {
+    BootstrapperImpl.enableFederation = originalFederationStatus;
+  }
+
+  // Assures that resource deletions happening in one control plane do not trigger deletion events
+  /// in watchers of resources on other control planes.
+  @Test
+  public void isolatedResourceDeletions() throws InterruptedException {
+    // Add the mock watcher for the normal server resource. The test control plane will send
+    // a deletion event.
+    xdsClient.watchXdsResource(XdsListenerResource.getInstance(), "test-server", mockWatcher);
+    verify(mockWatcher, timeout(20000)).onResourceDoesNotExist("test-server");
```

Comment:
onResourceDoesNotExist is triggered by a 15 second timeout. That will make this test very slow. It seems like there's more than just this delay, as the test takes a ridiculous 43.196s to run on my machine. (Line 89, for example)

Reply:
I cut down on all the unessential verify() calls - it's now down to 15 seconds...